packages/bvaughn-architecture-demo/components/console/MessagesList.tsx

```diff
@@ -0,0 +1,80 @@
+import { Message } from "@replayio/protocol";
+import { useContext } from "react";
+
+import { ConsoleFiltersContext, FocusContext, SessionContext } from "../../src/contexts";
+import useFilteredMessages from "../../src/hooks/useFilteredMessages";
+import { getMessages } from "../../src/suspense/MessagesCache";
+import { getClosestPointForTime } from "../../src/suspense/PointsCache";
+
+import MessageRenderer from "./MessageRenderer";
+import styles from "./MessagesList.module.css";
+
+// This is a crappy approximation of the console; the UI isn't meant to be the focus of this branch.
+// The primary purpose of this component is to showcase:
+// 1. The getMessages() Suspense cache for just-in-time loading of console data
+// 2. The memoized useFilteredMessages() selector for computing derived (merged, sorted, and filtered) data
+//
+// Note that the props passed from the parent component would more likely be exposed through Context in a real app.
+// We're passing them as props in this case since the parent and child are right beside each other in the tree.
+export default function MessagesList() {
+  const { sessionId } = useContext(SessionContext);
+
+  const { filterByText, levelFlags } = useContext(ConsoleFiltersContext);
+  const { range, isTransitionPending: isFocusTransitionPending } = useContext(FocusContext);
+
+  let focusMode = null;
+  if (range !== null) {
+    const [startTime, endTime] = range;
+
+    const startPoint = getClosestPointForTime(startTime, sessionId);
+    const endPoint = getClosestPointForTime(endTime, sessionId);
+
+    focusMode = {
+      begin: {
+        point: startPoint,
+        time: startTime,
+      },
+      end: {
+        point: endPoint,
+        time: endTime,
+      },
+    };
+  }
+
+  const { countAfter, countBefore, didOverflow, messages } = getMessages(sessionId, focusMode);
+
+  // Memoized selector that joins log points and messages and filters by criteria (e.g. type)
+  // Note that we are intentionally not storing derived values like this in state.
+  const filteredMessages = useFilteredMessages(messages, {
+    ...levelFlags,
+    filterByText,
```

Comment:
Here's another key part of the proposal: _Derived values_ (like filtered messages) are just memoized selectors. We don't need to store them in parallel "state" (or Redux).